docs/python/tutorials/upgrading-workflows.md

```diff
@@ -0,0 +1,114 @@
+Therefore, if `DBOS.patch()` is `True`, call the new code, else, call the old code.
+
+To use patching, you must enable it in configuration:
+
+```python
+config: DBOSConfig = {
+    "name": "dbos-app",
+    "system_database_url": os.environ.get("DBOS_SYSTEM_DATABASE_URL"),
+    "enable_patching": True,
+}
+DBOS(config=config)
+```
+
+For example, let's say our workflow is:
+
+```python
+@DBOS.workflow()
+def workflow():
+  foo()
+  bar()
+```
+
+We want to replace the call to `foo()` with a call to `baz()`.
+This is a breaking change because it changes what steps run.
+We can make this breaking change safely using a patch:
+
+```python
+@DBOS.workflow()
+def workflow():
+  if DBOS.patch("use-baz"):
+    baz()
+  else:
+    foo()
+  bar()
+```
+
+Now, new workflows will run `baz()`, while old workflows will safely continue through `foo()`.
+
+### Deprecating and Removing Patches
+
+Patches don't need to stay in your code forever.
+Once all workflows that started before you deployed the patch are complete, you can safely remove patches from your code.
+First, you must deprecate the patch with [`DBOS.deprecate_patch()`](../reference/contexts.md#deprecate_patch).
+This safely runs all workflows that contain the patch marker, but does not insert the patch marker into new workflows.
+For example, here's how to deprecate the patch above:
+
+```python
+@DBOS.workflow()
+def workflow():
+  DBOS.deprecate_patch("use-baz")
+  baz()
+  bar()
+```
+
+Then, when all workflows that started before you deprecated the patch are complete, you can remove the patch entirely:
+
+```python
+@DBOS.workflow()
+def workflow():
+  baz()
+  bar()
+```
+
+If any mistakes happen during the process (a breaking change is not patched, or a patch is deprecated or removed prematurely), the workflow will throw a `DBOSUnexpectedStepError` error clearly pointing to the step where the problem occurred.
+
+### How Patching Works
+
+Under the hood, when you call `DBOS.patch()` from a workflow, it attempts to insert a "patch marker" at its current point in your workflow history.
+If it succesfully inserts the patch marker or if the patch marker is already present, then the workflow must be new (it started after the patch, or started before the patch but has not yet reached this point), so `DBOS.patch()` returns `True`.
+If there is already a record present in this point in your workflow history, then the workflow must be old (it started before the patch and has already continued past this point), so `DBOS.patch()` returns `False`.
+
+When you deprecate a patch with `DBOS.deprecate_patch()`, new workflows no longer insert patch markers into their workflow history.
+However, if a workflow contains the patch marker in its history, it continues past that patch marker, safely ignoring it.
+Once all workflows with patch markers are complete, you can safely remove the patch entirely.
```

Comment:
Shall we add "you can use the DBOS Client list_workflows method [link] to see which workflows are in-flight?"